ODIN_II/SRC/OdinGridAnalyzer.cpp

```diff
@@ -0,0 +1,325 @@
+
+            //Fill in the region
+            for (size_t x = x_start; x + (type->width - 1) <= x_max; x += incrx) {
+                for (size_t y = y_start; y + (type->height - 1) <= y_max; y += incry) {
+                    set_grid_block_type(grid_loc_def->priority, type, x, y, grid);
+                }
+            }
+        }
+    }
+}
+/**
+ *  First is type (char*), second is priority
+ */
+void OdinGridAnalyzer::set_grid_block_type(int priority, t_physical_tile_type* type, int x_root, int y_root, std::vector<std::vector<std::pair<char*, int>>>& grid) {
+    if (grid[x_root][y_root].second < priority) {
+        grid[x_root][y_root].first = type->name;
+        grid[x_root][y_root].second = priority;
+
+        // This block is commented out, to simplify counting
+        // if this code is enabled, the hard block will occupy as
+        // much grid tiles as it should, however, a special counting
+        // procedure should be implemented. That being said,
+        // somebody might need that code in future, whereas
+        // it is not required at current state
+
+        // for (size_t x = x_root; x < x_root + type->width; ++x) {
+        //     size_t x_offset = x - x_root;
+        //     for (size_t y = y_root; y < y_root + type->height; ++y) {
+        //         size_t y_offset = y - y_root;
+        //         auto& grid_tile = grid[x][y];
+        //         grid[x][y].first = type->name;
+        //         grid[x][y].second = priority;
+        //     }
+        // }
+    }
+}
+
+int OdinGridAnalyzer::findDesiredBlockPriority(t_grid_def& layout, std::string& typeTag) {
+    int result = -1;
+    std::vector<t_grid_loc_def>::iterator pointer;
+    for (pointer = layout.loc_defs.begin(); pointer < layout.loc_defs.end(); pointer++) {
+        if (typeTag.compare(pointer->block_type) == 0) {
+            result = pointer->priority;
+        }
+    }
+    //TODO are there any unspecified priorities?
+    return result;
+}
+
+ssize_t OdinGridAnalyzer::findIndexOfAssociatedTileType(std::vector<t_physical_tile_type> tileTypes, std::string& typeTag) {
+    ssize_t result = -1;
+    ssize_t iter = -1;
+    std::cout << "The type tag is" << typeTag << std::endl;
+    std::vector<t_physical_tile_type>::iterator pointer;
+
+    for (pointer = tileTypes.begin(); pointer < tileTypes.end(); pointer++) {
+        iter++;
+        std::string stringForComparison = std::string(pointer->name);
+        std::cout << stringForComparison << std::endl;
+        if (stringForComparison.compare(typeTag) == 0) {
+            result = iter;
+            break;
+        }
+    }
+    if (result == -1) {
+        std::cerr << "Was not able to find associated tile type" << std::endl;
+    }
+    return result;
+}
+
+std::string OdinGridAnalyzer::getArchDescriptionTag(int hardBlockType) {
+    std::string result;
+    switch (hardBlockType) {
+        case mix_hard_blocks::MULTIPLIERS: {
+            result = "mult_36";
```

Comment:
not entirely true, multipliers can also be 24, 18, 9 etcc... theres limitless implementations, and also fracturable. I wonder if you could maybe find a better way here? 

also you could have more than 1 adder type (clb carry-chain vs dsp adder)

Reply:
I wonder if it wouldnt be best to forgo of the availability here, and just let somthing else crash if it doesnt exist

Reply:
Isn't that what default case currently does?
also, this one is quite complicated. There was no obvious way to do that besides adding the tags manually.

Reply:
oh I see, you are trying to map Odin's enum with vpr hard blocks here. I think this is just a flaw in the architecture and I don't think it's a bad idea to ask that a tag gets added to allow Odin to figure out the component usage. I think the architecture is just a bit limiting...